File: po/anime.m
%% 主程序（带实时动画）
clear; clc;

% 参数设置
N = 250;        % 种群规模（建议调小以提高动画流畅度）
dim = 2;        % 二维问题
Max_iter = 1500; % 迭代次数
lb = -10;       % 搜索范围
ub = 10;

% 预计算函数曲面
[X,Y] = meshgrid(linspace(lb,ub,2500));
Z = arrayfun(@(x,y) shubert([x y]), X, Y);

% 初始化图形窗口
figure('Position', [100 100 1200 600])
ax1 = subplot(1,2,1,'Parent',gcf);
surf(ax1, X,Y,Z,'EdgeColor','none','FaceAlpha',0.7);
hold(ax1, 'on');
scatter_handle = scatter3(ax1, 0,0,0, 30, 'r','filled');
title(ax1,'实时粒子搜索轨迹');
xlabel('x'); ylabel('y'); zlabel('f(x,y)');
view(ax1, 135, -8); % 原代码第19行替换为这个视角
colormap(ax1, parula) % 更改为parula色图增强低值区对比度
colorbar(ax1, 'southoutside') % 将色条移到下方避免遮挡

% 收敛曲线子图
ax2 = subplot(1,2,2);
conv_plot = animatedline(ax2, 'Color','b','LineWidth',1.5);
title(ax2,'收敛曲线');
xlabel('迭代次数'); ylabel('最优适应度');
grid on
axis([0 Max_iter -200 0])

% 运行优化算法（修改后的函数）
[best_pos, best_fit, Convergence_curve, all_positions] = parrot_optimizer_animated(N, Max_iter, lb, ub, dim);

% 生成动画视频（可选）
writerObj = VideoWriter('PO_Animation.mp4','MPEG-4');
open(writerObj);

% 实时动画显示
for iter = 1:Max_iter
    current_pos = all_positions{iter};
    fitness_vals = shubert(current_pos);
    
    % 更新粒子位置散点图
    set(scatter_handle, 'XData', current_pos(:,1),...
                       'YData', current_pos(:,2),...
                       'ZData', fitness_vals);
    
    % 更新收敛曲线
    addpoints(conv_plot, iter, Convergence_curve(iter));
    
    % 捕获视频帧（取消注释保存视频）
    frame = getframe(gcf);
    writeVideo(writerObj, frame);
    
    % 控制动画速度
    drawnow limitrate
    pause(0.05)
end

% 关闭视频写入（取消注释保存视频）
close(writerObj);

%% 修改后的PO算法（带轨迹记录）
function [X_best, f_best, Convergence_curve, all_positions] = parrot_optimizer_animated(N, Max_iter, lb, ub, dim)
    % 初始化
    X = initialization(N, dim, ub, lb);
    fitness = shubert(X);
    [f_best, idx] = min(fitness);
    X_best = X(idx,:);
    all_positions = cell(Max_iter,1);  % 新增轨迹存储
    
    % 迭代优化
    for t = 1:Max_iter
        X_mean = mean(X, 1);
        all_positions{t} = X;  % 记录当前代粒子位置
        
        for i = 1:N
            % 随机行为选择（同原算法）
            St = randi([1,4]);
            
            % 行为1：觅食行为
            if St == 1
                beta = 1.5;  % Levy指数
                sigma = (gamma(1+beta)*sin(pi*beta/2)/(gamma((1+beta)/2)*beta*2^((beta-1)/2)))^(1/beta);
                u = normrnd(0, sigma, [1,dim]);
                v = normrnd(0, 1, [1,dim]);
                Levy = u./(abs(v).^(1/beta));  % Levy飞行向量
                
                step = rand*(1 - t/Max_iter)^(2*t/Max_iter);
                X_new = (X(i,:) - X_best).*Levy + step*X_mean;
            
            % 行为2：停留行为    
            elseif St == 2
                beta = 1.5;
                sigma = (gamma(1+beta)*sin(pi*beta/2)/(gamma((1+beta)/2)*beta*2^((beta-1)/2)))^(1/beta);
                u = normrnd(0, sigma, [1,dim]);
                v = normrnd(0, 1, [1,dim]);
                Levy = u./(abs(v).^(1/beta));
                
                X_new = X(i,:) + X_best.*Levy + rand(1,dim);
            
            % 行为3：交流行为    
            elseif St == 3
                if rand <= 0.5
                    factor = 0.2*rand*(1 - t/Max_iter);
                    X_new = factor*(X(i,:) - X_mean);
                else
                    factor = 0.2 * rand(1,dim) .* exp(-t./(rand(1,dim)*Max_iter));
                    X_new = X(i,:) + factor;  % 添加随机扰动向量
                end
            
            % 行为4：恐惧行为    
            else
                theta = 0.8*pi*t/Max_iter;
                term1 = rand*cos(theta)*(X_best - X(i,:));
                term2 = cos(rand*pi)*(t/Max_iter)^(2/Max_iter)*(X(i,:) - X_best);
                X_new = X(i,:) + term1 - term2;
            end

            % 边界约束
            X_new = min(max(X_new, lb), ub);
            
            % 贪婪选择
            f_new = shubert(X_new);
            if f_new < fitness(i)
                X(i,:) = X_new;
                fitness(i) = f_new;
            end
        end
        
        % 更新全局最优
        [current_min, idx] = min(fitness);
        if current_min < f_best
            f_best = current_min;
            X_best = X(idx,:);
        end
        
        Convergence_curve(t) = f_best;
    end
end

%% Shubert目标函数
function y = shubert(x)
    % 输入：x为N×2矩阵，每行代表一个个体（二维）
    % 输出：N×1向量，每个个体的适应度值
    [N, ~] = size(x);
    y = zeros(N,1);
    for k = 1:N
        t1 = 0; t2 = 0;
        for i = 1:5
            t1 = t1 + i * cos((i+1)*x(k,1) + i);
            t2 = t2 + i * cos((i+1)*x(k,2) + i);
        end
        y(k) = t1 * t2;
    end
end
%% 种群初始化函数
function X = initialization(N, dim, ub, lb)
    % 生成均匀分布的初始种群
    X = rand(N, dim).*(ub - lb) + lb;
end

File: po/anime.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from scipy.special import gamma
import time


# 目标函数 Shubert
def shubert(x):
    t1 = np.zeros(x.shape[0])
    t2 = np.zeros(x.shape[0])
    for i in range(1, 6):
        t1 += i * np.cos((i + 1) * x[:, 0] + i)
        t2 += i * np.cos((i + 1) * x[:, 1] + i)
    return t1 * t2


# 种群初始化
def initialization(N, dim, lb, ub):
    return np.random.uniform(lb, ub, size=(N, dim))


# 改进的鹦鹉优化算法（带轨迹记录）
def parrot_optimizer_animated(N, Max_iter, lb, ub, dim):
    # 初始化
    X = initialization(N, dim, lb, ub)
    fitness = shubert(X)
    f_best = np.min(fitness)
    idx = np.argmin(fitness)
    X_best = X[idx, :]
    all_positions = []
    Convergence_curve = np.zeros(Max_iter)

    for t in range(Max_iter):
        X_mean = np.mean(X, axis=0)
        all_positions.append(X.copy())

        for i in range(N):
            St = np.random.randint(1, 5)  # 1-4

            if St == 1:  # 觅食行为
                beta = 1.5
                sigma = (gamma(1 + beta) * np.sin(np.pi * beta / 2) / (
                            gamma((1 + beta) / 2) * beta * 2 ** ((beta - 1) / 2))) ** (1 / beta)
                u = np.random.normal(0, sigma, dim)
                v = np.random.normal(0, 1, dim)
                Levy = u / np.abs(v) ** (1 / beta)
                step = np.random.rand() * (1 - (t + 1) / Max_iter) ** (2 * (t + 1) / Max_iter)
                X_new = (X[i] - X_best) * Levy + step * X_mean

            elif St == 2:  # 停留行为
                beta = 1.5
                sigma = (gamma(1 + beta) * np.sin(np.pi * beta / 2) / (
                            gamma((1 + beta) / 2) * beta * 2 ** ((beta - 1) / 2))) ** (1 / beta)
                u = np.random.normal(0, sigma, dim)
                v = np.random.normal(0, 1, dim)
                Levy = u / np.abs(v) ** (1 / beta)
                X_new = X[i] + X_best * Levy + np.random.rand(dim)

            elif St == 3:  # 交流行为
                if np.random.rand() <= 0.5:
                    factor = 0.2 * np.random.rand() * (1 - (t + 1) / Max_iter)
                    X_new = factor * (X[i] - X_mean)
                else:
                    factor = 0.2 * np.random.rand(dim) * np.exp(-(t + 1) / (np.random.rand(dim) * Max_iter))
                    X_new = X[i] + factor

            else:  # 恐惧行为
                theta = 0.5 * np.pi * (t + 1) / Max_iter
                term1 = np.random.rand() * np.cos(theta) * (X_best - X[i])
                term2 = np.cos(np.random.rand() * np.pi) * ((t + 1) / Max_iter) ** (2 / Max_iter) * (X[i] - X_best)
                X_new = X[i] + term1 - term2

            # 边界处理
            X_new = np.clip(X_new, lb, ub)

            # 贪婪选择
            f_new = shubert(X_new.reshape(1, -1))[0]
            if f_new < fitness[i]:
                X[i] = X_new
                fitness[i] = f_new

        # 更新全局最优
        current_min = np.min(fitness)
        if current_min < f_best:
            f_best = current_min
            idx = np.argmin(fitness)
            X_best = X[idx].copy()

        Convergence_curve[t] = f_best

    return X_best, f_best, Convergence_curve, all_positions


# 主程序
if __name__ == "__main__":
    # 参数设置
    N = 100  # 种群规模（适当减小以提高性能）
    Max_iter = 50  # 迭代次数
    lb, ub = -10, 10
    dim = 2

    # 预计算函数曲面（适当降低分辨率）
    x = np.linspace(lb, ub, 500)
    y = np.linspace(lb, ub, 500)
    X_grid, Y_grid = np.meshgrid(x, y)
    points = np.column_stack((X_grid.ravel(), Y_grid.ravel()))
    Z = shubert(points).reshape(X_grid.shape)

    # 初始化图形界面
    plt.ion()
    fig = plt.figure(figsize=(12, 6))

    # 3D曲面子图
    ax1 = fig.add_subplot(121, projection='3d')
    surf = ax1.plot_surface(X_grid, Y_grid, Z, cmap='jet',
                            rstride=2, cstride=2, alpha=0.6,
                            antialiased=False)
    scat = ax1.scatter([], [], [], s=30, c='r', alpha=0.8)
    ax1.set_title('Real-time Particle Trajectory')
    ax1.set_xlabel('x')
    ax1.set_ylabel('y')
    ax1.set_zlabel('f(x,y)')
    ax1.view_init(elev=30, azim=-45)
    fig.colorbar(surf, ax=ax1, shrink=0.6)

    # 收敛曲线子图
    ax2 = fig.add_subplot(122)
    line, = ax2.plot([], [], 'b-', linewidth=1.5)
    ax2.set_title('Convergence Curve')
    ax2.set_xlabel('Iteration')
    ax2.set_ylabel('Best Fitness')
    ax2.grid(True)
    ax2.set_xlim(0, Max_iter)
    ax2.set_ylim(-200, 0)

    # 运行优化算法
    start_time = time.time()
    X_best, best_fit, conv_curve, all_positions = parrot_optimizer_animated(
        N, Max_iter, lb, ub, dim)
    print(f"Optimization completed in {time.time() - start_time:.2f}s")

    # 动画更新
    for t in range(Max_iter):
        current_pos = all_positions[t]
        fitness_vals = shubert(current_pos)

        # 更新3D散点
        scat._offsets3d = (current_pos[:, 0], current_pos[:, 1], fitness_vals)

        # 更新收敛曲线
        line.set_data(np.arange(t + 1), conv_curve[:t + 1])

        # 调整视图
        fig.canvas.draw()
        fig.canvas.flush_events()

        # 控制动画速度
        plt.pause(0.05)

    plt.ioff()
    plt.show()